typed-racket-lib/typed-racket/rep/type-rep.rkt

```diff
@@ -227,12 +230,36 @@
 ;; Vectors
 ;;----------
 
-(def-type VectorTop () [#:mask mask:vector]
-  [#:singleton -VectorTop])
+(def-structural Immutable-Vector ([elem #:covariant])
+  [#:mask mask:immutable-vector])
 
-;; elem is a Type
-(def-structural Vector ([elem #:invariant])
-  [#:mask mask:vector])
+(def-type Mutable-VectorTop ()
+  [#:mask mask:mutable-vector]
+  [#:singleton -Mutable-VectorTop])
+
+(def-structural Mutable-Vector ([elem #:invariant])
+  [#:mask mask:mutable-vector])
+
+(define-match-expander Vector:
+  (lambda (stx)
+    (syntax-parse stx
+     [(_ elem-pat)
+      #'(or (Immutable-Vector: elem-pat)
+            (Mutable-Vector: elem-pat)
+            ;; Cannot use `list-no-order`
```

Comment:
You could use `list-no-order` with an explicit equality check.

Reply:
Like this, or is there a simpler way?
```
(and (Union-all: (list-no-order (.... elem-pat) (.... elem-pat2)))
     (? (lambda (_) (equal? elem-pat elem-pat2))))
```